File: src/dannsub.f
C Output from Public domain Ratfor, version 1.01
      subroutine dannsub(n,p,x,y,class, x0,kmetric,iter,fullw,scalar, ep
     *silon,which,dist,covw,covmin,means,weight,covb,scratch)
      implicit double precision (a-h,o-z)
      integer n, p, class, kmetric, which(n), y(n),iter
      double precision x(n,p), weight(n), means(class,p),dist(n),scratch
     *(n), covw(p,p),x0(p),covb(p,p)
      double precision epsilon, wmax,covmin
      logical fullw,singul,scalar
      scalar=.true.
      fullw=.false.
      icv=0
      do23000 j=1,p
      do23002 k=1,p
      covb(j,k)=0d0
23002 continue
23003 continue
23000 continue
23001 continue
      do23004 i=1,n 
      do23006 j=1,p
      x0(j)=x(i,j)
23006 continue
23007 continue
      call nndist(n,p,class,kmetric,x,x0,icv,y,iter, fullw,scalar,epsilo
     *n,which,dist,covw,covmin,means, weight,scratch,scratch(p+1),scratc
     *h(p*p+p+1), scratch(2*p*p+p+1))
      do23008 j=1,p
      do23010 k=1,p
      tot=0d0
      do23012 jk=1,class
      tot=tot+means(jk,j)*means(jk,k)
23012 continue
23013 continue
      covb(j,k)=covb(j,k)+tot/n
23010 continue
23011 continue
23008 continue
23009 continue
23004 continue
23005 continue
      return
      end

File: R/dannsub.R
#' dannsub
#'
#' @param x covariate matrix
#' @param km max subspace
#' @param y  labels
#' @param k  number of clusters
#' @param epsilon error
#' @param fullw  Boolean
#' @param scalex Boolean, default true
#' @param scalar Boolean, 
#' @param iter   Max number of iterations
#' @param iter.sub Max. number of subspace iterations
#' @param method  knn
#' @param dims dims
#' @param xnew new x
#' @param ynew new ys
#'
#' @return an object of class dannsub
#'
#' @examples \dontrun{
#' dannsub(km = max(50, 0.2 * n), k = 5, epsilon = 1, fullw = FALSE, scalex = TRUE, 
#'    scalar = TRUE, dims = seq(p), iter.sub = 1, iter = 1, method = "knn", xnew, ynew)
#' }

To fix the code, I made several adjustments and improvements. Here's the corrected version of the dannsub function:

R
Copy code
dannsub <- function(x, y, km = max(50, 0.2 * n), k = 5, epsilon = 1, fullw = FALSE, scalex = TRUE, scalar = TRUE, dims = seq(p), iter.sub = 1, iter = 1, method = "knn", xnew, ynew) {
  if (scalex) {
    mm <- apply(x, 2, mean)
    dd <- sqrt(apply(x, 2, var))
    x <- scale(x, center = mm, scale = dd)
  }
  
  storage.mode(x) <- "double"
  storage.mode(y) <- "integer"
  nclass <- length(table(y))
  y <- as.integer(factor(y))
  np <- dim(x)
  p <- np[2]
  n <- np[1]
  
  errors <- matrix(0, length(dims), length(epsilon))
  dimnames(errors) <- list(format(round(dims, 0)), format(round(epsilon, 1)))
  
  between <- .Fortran("dannsub", as.integer(np[1]), as.integer(np[2]), x, y, as.integer(nclass), double(np[2]), as.integer(km), as.integer(iter.sub), fullw, scalar, epsilon = as.double(epsilon[1]), which = integer(n), dist = double(n), covw = matrix(double(p^2), p, p), covmin = as.double(1e-04), means = matrix(double(nclass * p), nclass, p), weight = double(n), between = matrix(double(p^2), p, p), double(n + 2 * p^2 + 3 * p), PACKAGE = "dann")$between
  
  eb <- eigen(between)
  x <- x %*% eb$vectors
  
  if (missing(xnew)) {
    if (method == "knn") {
      for (i in seq_along(dims)) {
        errors[i, ] <- attr(knncv(x[, seq(dims[i]), drop = FALSE], y, k), "error")
      }
    } else {
      for (i in seq_along(dims)) {
        errors[i, ] <- apply(y != dann(x[, seq(dims[i]), drop = FALSE], y = y, k = k, kmetric = km, epsilon = epsilon, cv = TRUE, iter = iter), 2, sum)
      }
    }
  } else {
    cat("# New x and y data for testing\n")
    
    if (scalex) {
      xnew <- scale(xnew, center = mm, scale = dd)
    }
    
    xnew <- xnew %*% eb$vectors
    
    if (method == "knn") {
      for (i in seq_along(dims)) {
        errors[i, ] <- sum(knn(x[, seq(dims[i]), drop = FALSE], xnew[, seq(dims[i]), drop = FALSE], y, k) != ynew)
      }
    } else {
      for (i in seq_along(dims)) {
        errors[i, ] <- apply(ynew != dann(x[, seq(dims[i]), drop = FALSE], xnew[, seq(dims[i]), drop = FALSE], y = y, k = k, kmetric = km, epsilon = epsilon, cv = FALSE, iter = iter), 2, sum)
      }
    }
  }
  
  eb$errors <- errors
  
  if (scalex) {
    eb$scale <- dd
    eb$centers <- mm
  }
  
  class(eb) <- "dannsub"
  eb
}
